tokio/tests/task_hooks.rs

```diff
@@ -73,3 +73,113 @@ fn terminate_task_hook_fires() {
+    let spawns = Arc::new(AtomicUsize::new(0));
+    let poll_starts = Arc::new(AtomicUsize::new(0));
+    let poll_ends = Arc::new(AtomicUsize::new(0));
+
+    let runtime = Builder::new_current_thread()
+        .on_task_spawn(mk_spawn_location_hook(
+            "(current_thread) on_task_spawn",
+            &spawns,
+        ))
+        .on_before_task_poll(mk_spawn_location_hook(
+            "(current_thread) on_before_task_poll",
+            &poll_starts,
+        ))
+        .on_after_task_poll(mk_spawn_location_hook(
+            "(current_thread) on_after_task_poll",
+            &poll_ends,
+        ))
+        .build()
+        .unwrap();
+
+    let task = runtime.spawn(async move { tokio::task::yield_now().await });
+    runtime.block_on(async move {
+        task.await.unwrap();
+        // tick the runtime a bunch to close out tasks
+        for _ in 0..ITERATIONS {
+            tokio::task::yield_now().await;
+        }
+    });
+
+    assert_eq!(spawns.load(Ordering::SeqCst), 1);
+    let poll_starts = poll_starts.load(Ordering::SeqCst);
+    assert!(poll_starts > 1);
+    assert_eq!(poll_starts, poll_ends.load(Ordering::SeqCst));
+}
+
+/// Test that the correct spawn location is provided to the task hooks on a
+/// multi-thread runtime.
+///
+/// Testing this separately is necessary as the spawn code paths are different
+/// and we should ensure that `#[track_caller]` is passed through correctly
+/// for both runtimes.
+#[cfg_attr(
+    target_os = "wasi",
+    ignore = "WASI does not support multi-threaded runtime"
+)]
+#[test]
+fn task_hook_spawn_location_multi_thread() {
+    let spawns = Arc::new(AtomicUsize::new(0));
+    let poll_starts = Arc::new(AtomicUsize::new(0));
+    let poll_ends = Arc::new(AtomicUsize::new(0));
+
+    let runtime = Builder::new_multi_thread()
+        .on_task_spawn(mk_spawn_location_hook(
+            "(multi_thread) on_task_spawn",
+            &spawns,
+        ))
+        .on_before_task_poll(mk_spawn_location_hook(
+            "(multi_thread) on_before_task_poll",
+            &poll_starts,
+        ))
+        .on_after_task_poll(mk_spawn_location_hook(
+            "(multi_thread) on_after_task_poll",
+            &poll_ends,
+        ))
+        .build()
+        .unwrap();
+
+    let task = runtime.spawn(async move { tokio::task::yield_now().await });
+    runtime.block_on(async move {
+        task.await.unwrap();
+
+        // tick the runtime a bunch to close out tasks
+        for _ in 0..ITERATIONS {
+            tokio::task::yield_now().await;
+        }
+    });
+
+    // Give the runtime to shut down so that we see all the expected calls to
+    // the task hooks.
+    runtime.shutdown_timeout(std::time::Duration::from_secs(60));
+
+    assert_eq!(spawns.fetch_add(0, Ordering::SeqCst), 1);
+    let poll_starts = poll_starts.fetch_add(0, Ordering::SeqCst);
+    assert!(poll_starts > 1);
+    assert_eq!(poll_starts, poll_ends.fetch_add(0, Ordering::SeqCst));
```

Comment:
Not a blocker.

I believe `fetch_add` is worth a comment to explain the "why".